File: src/modules/07-smpp-integration/smpp-client.ts
import { loadConfig, type ParsedEnvConfig } from '../../config/env.js';
import { traceStore } from '../../tracing/trace-store.js';
import type { SmsMessage, SubmissionResult } from '../../types/sms.js';
import { getLogger } from '../../utils/logger.js';
import { toSmppValidityPeriod } from '../08-smpp-validity/validity-period.js';
import type { DeliverSmPdu, Session, SubmitSmOptions } from 'smpp';

const logger = getLogger();

/**
 * Real SMSC integration via SMPP — requirement #7.
 *
 * A real SMPP 3.4 client (npm `smpp`, farhadi/node-smpp) with bind lifecycle,
 * submit_sm construction using the real fields from modules 08/09 (validity
 * period, priority flag), automatic reconnect and enquire_link keepalive.
 *
 * Mirrors the HttpTowerSource pattern: fully implemented against the real
 * protocol, but every connection method throws a clear "awaiting credentials"
 * error when SMPP_HOST/SMPP_SYSTEM_ID are not configured — it never fabricates
 * a working connection.
 */

export class SmppClient {
  private session: Session | null = null;
  private reconnectTimer: NodeJS.Timeout | null = null;
  private closedByUs = false;
  private readonly cfg: ParsedEnvConfig;

  constructor(cfg: ParsedEnvConfig = loadConfig()) {
    this.cfg = cfg;
  }

  /** Whether real SMSC credentials exist to bind against. */
  isConfigured(): boolean {
    return Boolean(this.cfg.SMPP_HOST && this.cfg.SMPP_SYSTEM_ID);
  }

  /**
   * Connect and bind. Throws a clear error until real C-DOT credentials are
   * provided — never a fake session.
   */
  async connect(): Promise<Session> {
    if (!this.isConfigured()) {
      throw new Error(
        'SMPP credentials not configured (SMPP_HOST / SMPP_SYSTEM_ID). Awaiting C-DOT SMSC sandbox credentials.',
      );
    }
    if (this.session) return this.session;

    logger.info({ host: this.cfg.SMPP_HOST, port: this.cfg.SMPP_PORT }, 'smpp.connect');
    const session = (await import('smpp')).default.connect({
      host: this.cfg.SMPP_HOST,
      port: this.cfg.SMPP_PORT,
      connectTimeout: this.cfg.SMPP_SUBMIT_TIMEOUT_MS,
      auto_enquire_link_period: this.cfg.SMPP_ENQUIRE_LINK_PERIOD_MS / 1000,
    });

    await new Promise<void>((resolve, reject) => {
      const bindOpts = {
        system_id: this.cfg.SMPP_SYSTEM_ID,
        password: this.cfg.SMPP_PASSWORD ?? '',
        system_type: this.cfg.SMPP_SYSTEM_TYPE ?? '',
        interface_version: this.cfg.SMPP_INTERFACE_VERSION_NUM,
      };
      const bind = this.cfg.SMPP_BIND_MODE === 'transmitter' ? 'bind_transmitter' : 'bind_transceiver';
      session[bind](bindOpts, (pdu) => {
        if (pdu.command_status === 0) resolve();
        else reject(new Error(`SMPP bind failed with command_status=${pdu.command_status}`));
      });
    });

    this.session = session;
    session.on('close', () => this.handleDisconnect());
    session.on('error', (err) => logger.error({ err }, 'smpp.session_error'));
    logger.info('smpp.bound');
    return session;
  }

  /** Register a delivery-receipt callback (module 11 attaches here). */
  onDeliveryReceipt(cb: (pdu: DeliverSmPdu) => void): void {
    if (!this.session) throw new Error('No active SMPP session; connect() first');
    this.session.on('deliver_sm', (pdu) => cb(pdu as DeliverSmPdu));
  }

  /** Submit one message; resolves to a real SubmissionResult. */
  async submitSingle(message: SmsMessage): Promise<SubmissionResult> {
    const session = await this.connect();
    return this.submitOne(session, message);
  }

  /**
   * Submit a batch with a bounded in-flight window. Marks stage t3 on the
   * shared latency trace once the whole batch's submission completes.
   */
  async submitBatch(messages: readonly SmsMessage[], traceKey?: string): Promise<SubmissionResult[]> {
    const session = await this.connect();
    const results = new Array<SubmissionResult>(messages.length);
    let cursor = 0;

    const worker = async () => {
      while (cursor < messages.length) {
        const idx = cursor++;
        results[idx] = await this.submitOne(session, messages[idx]!);
      }
    };
    const workers = Array.from({ length: Math.min(this.cfg.SMPP_SUBMIT_CONCURRENCY, messages.length) }, () => worker());
    await Promise.all(workers);

    if (traceKey) {
      await traceStore.mark(traceKey, 't3', 'smpp.submit_complete', Date.now());
    }
    return results;
  }

  /** Close the session (cancels pending reconnect). */
  async close(): Promise<void> {
    this.closedByUs = true;
    if (this.reconnectTimer) clearTimeout(this.reconnectTimer);
    this.reconnectTimer = null;
    if (this.session) {
      const session = this.session;
      this.session = null;
      await new Promise<void>((resolve) => session.close(() => resolve()));
    }
  }

  private submitOne(session: Session, message: SmsMessage): Promise<SubmissionResult> {
    const pdu = buildSubmitSmPdu(message, this.cfg);
    return new Promise((resolve) => {
      const timer = setTimeout(() => {
        resolve({
          messageId: message.messageId,
          msisdn: message.msisdn,
          outcome: 'failed',
          errorText: `submit_sm timed out after ${this.cfg.SMPP_SUBMIT_TIMEOUT_MS}ms`,
        });
      }, this.cfg.SMPP_SUBMIT_TIMEOUT_MS);

      session.submit_sm(pdu, (resp) => {
        clearTimeout(timer);
        if (resp.command_status === 0) {
          resolve({
            messageId: message.messageId,
            msisdn: message.msisdn,
            outcome: 'accepted',
            smscMessageId: typeof resp.message_id === 'string' ? resp.message_id : undefined,
          });
        } else {
          resolve({
            messageId: message.messageId,
            msisdn: message.msisdn,
            outcome: 'rejected',
            errorCode: resp.command_status,
            errorText: smppErrorText(resp.command_status),
          });
        }
      });
    });
  }

  private handleDisconnect(): void {
    if (this.closedByUs) return;
    logger.warn({ delayMs: this.cfg.SMPP_RECONNECT_DELAY_MS }, 'smpp.disconnected_scheduling_reconnect');
    this.session = null;
    this.reconnectTimer = setTimeout(() => {
      this.reconnectTimer = null;
      void this.connect().catch((err) => logger.error({ err }, 'smpp.reconnect_failed'));
    }, this.cfg.SMPP_RECONNECT_DELAY_MS);
  }
}

/**
 * Build a real submit_sm PDU from a TURANT message + env config (pure — unit
 * tested without a live SMSC). Uses module 08 (validity_period) and module 09
 * (priority_flag) values already on the message. Validity is encoded in UTC
 * (nn = 00) with seconds truncated — the least ambiguous form for an SMSC.
 */
export function buildSubmitSmPdu(message: SmsMessage, cfg: ParsedEnvConfig = loadConfig()): SubmitSmOptions {
  const shortMessage = encodeMessageContent(message.content, message.dataCoding);
  return {
    source_addr_ton: cfg.SMPP_SRC_ADDR_TON,
    source_addr_npi: cfg.SMPP_SRC_ADDR_NPI,
    source_addr: cfg.SMPP_SRC_ADDR ?? '',
    dest_addr_ton: cfg.SMPP_DEST_ADDR_TON,
    dest_addr_npi: cfg.SMPP_DEST_ADDR_NPI,
    destination_addr: message.msisdn,
    esm_class: 0,
    protocol_id: 0,
    priority_flag: message.priorityFlag,
    schedule_delivery_time: undefined,
    validity_period: message.validityPeriod
      ? toSmppValidityPeriod(message.validityPeriod, { timezoneOffsetMinutes: 0, truncateSeconds: true })
      : undefined,
    registered_delivery: message.registeredDelivery,
    replace_if_present_flag: 0,
    data_coding: message.dataCoding === 'ucs2' ? 8 : 0,
    sm_default_msg_id: 0,
    short_message: shortMessage,
  };
}

/** Validate content length for the target coding; fail loudly, never truncate. */
export function encodeMessageContent(content: string, dataCoding: '7bit' | 'ucs2'): string {
  const maxChars = dataCoding === 'ucs2' ? 70 : 160;
  if (content.length === 0) throw new Error('SMS content must not be empty');
  if (content.length > maxChars) {
    throw new Error(`SMS content is ${content.length} characters; exceeds ${maxChars} for ${dataCoding} coding. Early-warning messages must fit one SMS.`);
  }
  return content;
}

const COMMAND_STATUS: Record<number, string> = {
  0: 'ESME_ROK',
  4: 'ESME_RINVMSGLEN',
  5: 'ESME_RINVCMDLEN',
  8: 'ESME_RINVSRCADR',
  0xa: 'ESME_RINVDSTADR',
  0xb: 'ESME_RINVMSGID',
  0xc: 'ESME_RBINDFAIL',
  0xd: 'ESME_RINVPASWD',
  0xe: 'ESME_RINVSYSID',
  0x11: 'ESME_RINVSRCTON',
  0x12: 'ESME_RINVSRCNPI',
  0x13: 'ESME_RINVDSTTON',
  0x14: 'ESME_RINVDSTNPI',
  0x400: 'ESME_RTHROTTLED',
  0x401: 'ESME_RINVSCHED',
  0x402: 'ESME_RINVEXPIRY',
  0x403: 'ESME_RINVDFTMSGID',
  0x404: 'ESME_RX_T_APPN',
  0x405: 'ESME_RX_P_APPN',
  0x406: 'ESME_RX_R_APPN',
  0x408: 'ESME_RQUERYFAIL',
  0x409: 'ESME_RINVOPTPARSTREAM',
};

export function smppErrorText(commandStatus: number | undefined): string {
  if (commandStatus === undefined) return 'command_status not present';
  return COMMAND_STATUS[commandStatus] ?? `command_status=${commandStatus}`;
}

File: src/main/java/com/turant/smpp/SmppClient.java
package com.turant.smpp;

import com.turant.dlr.DlrListener;
import com.turant.types.sms.DeliveryOutcome;
import com.turant.types.sms.SmsDataCoding;
import com.turant.types.sms.SmsMessage;
import com.turant.types.sms.SubmissionResult;
import org.jsmpp.InvalidResponseException;
import org.jsmpp.bean.*;
import org.jsmpp.extra.NegativeResponseException;
import org.jsmpp.extra.ProcessRequestException;
import org.jsmpp.extra.ResponseTimeoutException;
import org.jsmpp.session.BindParameter;
import org.jsmpp.session.DataSmResult;
import org.jsmpp.session.MessageReceiverListener;
import org.jsmpp.session.SMPPSession;
import org.jsmpp.session.SubmitSmResult;
import org.jsmpp.util.AbsoluteTimeFormatter;
import org.jsmpp.util.TimeFormatter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Real SMSC integration via SMPP - requirement #7.
 * 
 * A real SMPP 3.4 client (jSMPP) with bind lifecycle, submit_sm construction
 * using real fields from modules 08/09 (validity period, priority flag),
 * automatic reconnect and enquire_link keepalive.
 * 
 * Migrated from TypeScript Module 07 smpp-client.ts
 */
@Component
public class SmppClient {
    
    private static final Logger logger = LoggerFactory.getLogger(SmppClient.class);
    private static final TimeFormatter TIME_FORMATTER = new AbsoluteTimeFormatter();
    
    private SMPPSession session;
    private boolean closedByUs = false;
    private final ExecutorService executor;
    private final DlrListener dlrListener;
    
    @Value("${smpp.host:}")
    private String smppHost;
    
    @Value("${smpp.port:2775}")
    private int smppPort;
    
    @Value("${smpp.system-id:}")
    private String systemId;
    
    @Value("${smpp.password:}")
    private String password;
    
    @Value("${smpp.system-type:}")
    private String systemType;
    
    @Value("${smpp.bind-mode:transceiver}")
    private String bindMode;
    
    @Value("${smpp.interface-version:52}")
    private int interfaceVersion;
    
    @Value("${smpp.src-addr-ton:0}")
    private byte srcAddrTon;
    
    @Value("${smpp.src-addr-npi:0}")
    private byte srcAddrNpi;
    
    @Value("${smpp.src-addr:}")
    private String srcAddr;
    
    @Value("${smpp.dest-addr-ton:1}")
    private byte destAddrTon;
    
    @Value("${smpp.dest-addr-npi:1}")
    private byte destAddrNpi;
    
    @Value("${smpp.submit-timeout-ms:60000}")
    private long submitTimeoutMs;
    
    @Value("${smpp.submit-concurrency:10}")
    private int submitConcurrency;
    
    @Value("${smpp.reconnect-delay-ms:5000}")
    private long reconnectDelayMs;
    
    @Value("${smpp.enquire-link-period-ms:30000}")
    private long enquireLinkPeriodMs;
    
    public SmppClient(@Autowired(required = false) DlrListener dlrListener) {
        this.executor = Executors.newFixedThreadPool(16);
        this.dlrListener = dlrListener;
    }
    
    /**
     * Whether real SMSC credentials exist to bind against.
     */
    public boolean isConfigured() {
        return smppHost != null && !smppHost.isEmpty() && 
               systemId != null && !systemId.isEmpty();
    }
    
    /**
     * Connect and bind. Throws a clear error until real C-DOT credentials are provided.
     */
    public synchronized CompletableFuture<Void> connect() {
        if (!isConfigured()) {
            return CompletableFuture.failedFuture(
                new IllegalStateException(
                    "SMPP credentials not configured (SMPP_HOST / SMPP_SYSTEM_ID). " +
                    "Awaiting C-DOT SMSC sandbox credentials."
                )
            );
        }
        
        if (session != null && session.getSessionState().isBound()) {
            return CompletableFuture.completedFuture(null);
        }
        
        return CompletableFuture.supplyAsync(() -> {
            try {
                logger.info("Connecting to SMSC: host={}, port={}", smppHost, smppPort);
                
                session = new SMPPSession();
                session.setEnquireLinkTimer((int) (enquireLinkPeriodMs / 1000));
                session.setTransactionTimer(submitTimeoutMs);
                // Wire DLR listener to every new session (reconnect preserves correlation)
                if (dlrListener != null) {
                    session.setMessageReceiverListener(createDlrMessageReceiverListener());
                    logger.debug("DLR MessageReceiverListener registered on SMPPSession");
                } else {
                    logger.debug("No DlrListener available — deliver_sm will not be correlated");
                }
                
                BindParameter bindParam = new BindParameter(
                    BindType.BIND_TRX, // Use transceiver by default
                    systemId,
                    password,
                    systemType,
                    TypeOfNumber.valueOf(srcAddrTon),
                    NumberingPlanIndicator.valueOf(srcAddrNpi),
                    null
                );
                
                session.connectAndBind(smppHost, smppPort, bindParam);
                
                logger.info("SMPP session bound successfully");
                return null;
                
            } catch (IOException e) {
                logger.error("Failed to connect/bind to SMSC", e);
                throw new RuntimeException("SMPP connection failed", e);
            }
        }, executor);
    }
    
    /**
     * Submit one message; resolves to a real SubmissionResult.
     */
    public CompletableFuture<SubmissionResult> submitSingle(SmsMessage message) {
        return connect().thenCompose(v -> 
            CompletableFuture.supplyAsync(() -> submitOne(message), executor)
        );
    }
    
    /**
     * Submit a batch with bounded concurrency.
     */
    public CompletableFuture<List<SubmissionResult>> submitBatch(
            List<SmsMessage> messages, 
            String traceKey) {
        
        return connect().thenCompose(v -> {
            List<CompletableFuture<SubmissionResult>> futures = messages.stream()
                .map(msg -> CompletableFuture.supplyAsync(() -> submitOne(msg), executor))
                .toList();
            
            return CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
                .thenApply(ignored -> {
                    List<SubmissionResult> results = new ArrayList<>();
                    for (CompletableFuture<SubmissionResult> future : futures) {
                        results.add(future.join());
                    }
                    
                    logger.info("Batch submission completed: messages={}, traceKey={}", 
                        messages.size(), traceKey);
                    
                    // TODO: Mark t3 on trace store when implemented
                    // traceStore.mark(traceKey, "t3", "smpp.submit_complete", System.currentTimeMillis());
                    
                    return results;
                });
        });
    }
    
    /**
     * Close the session.
     */
    public void close() {
        closedByUs = true;
        if (session != null) {
            session.unbindAndClose();
            session = null;
        }
    }
    
    /**
     * Submit one message to SMSC.
     */
    private SubmissionResult submitOne(SmsMessage message) {
        try {
            if (session == null || !session.getSessionState().isBound()) {
                return new SubmissionResult(
                    message.messageId(),
                    message.msisdn(),
                    DeliveryOutcome.failed,
                    null,
                    null,
                    "SMPP session not bound",
                    null
                );
            }
            
            // Build submit_sm PDU
            String content = encodeMessageContent(message.content(), message.dataCoding());
            byte[] messageBytes = content.getBytes(
                message.dataCoding() == SmsDataCoding.UCS2 ? "UTF-16BE" : "ISO-8859-1"
            );
            
            // Use appropriate jSMPP Alphabet enum for data coding
            Alphabet alphabet = message.dataCoding() == SmsDataCoding.UCS2 
                ? Alphabet.ALPHA_UCS2 
                : Alphabet.ALPHA_DEFAULT;
            
            // Format validity period if present (Module 08)
            String validityPeriodStr = null;
            if (message.validityPeriod() != null) {
                validityPeriodStr = ValidityPeriod.toSmppValidityPeriod(
                    message.validityPeriod(), 
                    true // truncate seconds
                );
            }
            
            // Submit message with priority flag (Module 09) and validity period (Module 08)
            SubmitSmResult result = session.submitShortMessage(
                "CMT",
                TypeOfNumber.valueOf(srcAddrTon),
                NumberingPlanIndicator.valueOf(srcAddrNpi),
                srcAddr != null ? srcAddr : "",
                TypeOfNumber.valueOf(destAddrTon),
                NumberingPlanIndicator.valueOf(destAddrNpi),
                message.msisdn(),
                new ESMClass(),
                (byte) 0, // protocol_id
                message.priorityFlag(), // priority_flag from message (Module 09)
                null, // schedule_delivery_time
                validityPeriodStr, // validity_period from message.validityPeriod (Module 08)
                new RegisteredDelivery((byte) message.registeredDelivery()),
                (byte) 0, // replace_if_present_flag
                new GeneralDataCoding(alphabet),
                (byte) 0, // sm_default_msg_id
                messageBytes
            );
            
            String messageId = result.getMessageId();
            
            logger.debug("Message submitted: messageId={}, msisdn={}, smscMessageId={}", 
                message.messageId(), message.msisdn(), messageId);
            // Register for DLR correlation (existing DlrListener logic)
            if (dlrListener != null && messageId != null && !messageId.isBlank()) {
                try {
                    dlrListener.registerSubmission(messageId, message.messageId(), message.alertId(), message.msisdn());
                    logger.debug("DLR correlation registered: smscMessageId={}, alertId={}, msisdn={}", messageId, message.alertId(), message.msisdn());
                } catch (Exception e) {
                    logger.warn("Failed to register DLR correlation for smscMessageId={}", messageId, e);
                }
            }
            
            return new SubmissionResult(
                message.messageId(),
                message.msisdn(),
                DeliveryOutcome.accepted,
                messageId,
                null,
                null,
                null
            );
            
        } catch (ResponseTimeoutException | InvalidResponseException | IOException e) {
            logger.error("Failed to submit message: messageId=" + message.messageId(), e);
            return new SubmissionResult(
                message.messageId(),
                message.msisdn(),
                DeliveryOutcome.failed,
                null,
                null,
                e.getMessage(),
                null
            );
        } catch (NegativeResponseException e) {
            logger.warn("Message rejected by SMSC: messageId={}, commandStatus={}", 
                message.messageId(), e.getCommandStatus());
            return new SubmissionResult(
                message.messageId(),
                message.msisdn(),
                DeliveryOutcome.rejected,
                null,
                e.getCommandStatus(),
                getSmppErrorText(e.getCommandStatus()),
                null
            );
        } catch (Exception e) {
            // Catch any other unexpected exceptions
            logger.error("Unexpected error submitting message: messageId=" + message.messageId(), e);
            return new SubmissionResult(
                message.messageId(),
                message.msisdn(),
                DeliveryOutcome.failed,
                null,
                null,
                e.getMessage(),
                null
            );
        }
    }
    
    /**
     * Validate content length for the target coding; fail loudly, never truncate.
     */
    private String encodeMessageContent(String content, SmsDataCoding dataCoding) {
        int maxChars = dataCoding == SmsDataCoding.UCS2 ? 70 : 160;
        
        if (content.isEmpty()) {
            throw new IllegalArgumentException("SMS content must not be empty");
        }
        
        if (content.length() > maxChars) {
            throw new IllegalArgumentException(
                String.format("SMS content is %d characters; exceeds %d for %s coding. " +
                    "Early-warning messages must fit one SMS.", 
                    content.length(), maxChars, dataCoding.getValue())
            );
        }
        
        return content;
    }
    
    /**
     * Create MessageReceiverListener that routes deliver_sm to existing DlrListener.
     * Attached to every new SMPPSession (reconnect preserves correlation).
     * Handles DELIVRD, EXPIRED, UNDELIV, REJECTD, DELETED and extensible others.
     */
    private MessageReceiverListener createDlrMessageReceiverListener() {
        return new MessageReceiverListener() {
            @Override
            public void onAcceptDeliverSm(DeliverSm deliverSm) throws ProcessRequestException {
                try {
                    byte[] shortMsg = deliverSm.getShortMessage();
                    String receiptText = null;
                    if (shortMsg != null && shortMsg.length > 0) {
                        // jSMPP delivers short_message as bytes; SMPPSim uses GSM7 plain text
                        try {
                            receiptText = new String(shortMsg, java.nio.charset.StandardCharsets.UTF_8);
                            // If contains non-UTF8, fallback to GSM
                            if (receiptText.contains("\u0000")) {
                                receiptText = new String(shortMsg, java.nio.charset.StandardCharsets.ISO_8859_1);
                            }
                        } catch (Exception e) {
                            receiptText = new String(shortMsg, java.nio.charset.StandardCharsets.ISO_8859_1);
                        }
                    }
                    // Also check message_payload optional param if short_message empty
                    if ((receiptText == null || receiptText.isBlank()) && deliverSm.getOptionalParameters() != null) {
                        for (org.jsmpp.bean.OptionalParameter op : deliverSm.getOptionalParameters()) {
                            if (op instanceof org.jsmpp.bean.OptionalParameter.Message_payload) {
                                byte[] payload = ((org.jsmpp.bean.OptionalParameter.Message_payload) op).getValue();
                                if (payload != null) receiptText = new String(payload, java.nio.charset.StandardCharsets.UTF_8);
                            }
                        }
                    }
                    if (receiptText == null) receiptText = "";
                    // Trim for logging (never log credentials — receipt has no secrets)
                    String logText = receiptText.length() > 300 ? receiptText.substring(0, 300) + "..." : receiptText;
                    logger.debug("Received deliver_sm: {}", logText);

                    if (dlrListener != null) {
                        var parsed = DlrListener.parseDeliveryReceipt(receiptText);
                        if (parsed != null) {
                            String state = parsed.messageState();
                            // Handle expected states, but extensible — any stat is recorded
                            if (state != null) {
                                switch (state) {
                                    case "DELIVRD":
                                    case "EXPIRED":
                                    case "UNDELIV":
                                    case "REJECTD":
                                    case "DELETED":
                                    case "ACCEPTD":
                                    case "ENROUTE":
                                    case "UNKNOWN":
                                        logger.info("DLR deliver_sm parsed: smscMessageId={}, state={}, err={}", parsed.smscMessageId(), state, parsed.errorCode());
                                        break;
                                    default:
                                        logger.info("DLR deliver_sm with unlisted state: smscMessageId={}, state={}, err={}", parsed.smscMessageId(), state, parsed.errorCode());
                                }
                            }
                            // Correlate via existing logic (logs matched vs uncorrelated)
                            dlrListener.handleReceipt(receiptText, null);
                        } else {
                            logger.debug("Non-DLR deliver_sm ignored (no id/stat): {}", logText);
                            // Still try to handle via DlrListener for uncorrelated logging
                            dlrListener.handleReceipt(receiptText, null);
                        }
                    } else {
                        logger.warn("DLR received but no DlrListener registered — receipt not correlated");
                    }
                } catch (Exception e) {
                    logger.warn("Failed to process deliver_sm", e);
                    // Do not throw ProcessRequestException — we want to ACK the deliver_sm (ESME_ROK) even if parsing fails
                }
            }

            @Override
            public void onAcceptAlertNotification(AlertNotification alertNotification) {
                logger.debug("Received AlertNotification: {}", alertNotification);
            }

            @Override
            public DataSmResult onAcceptDataSm(DataSm dataSm, org.jsmpp.session.Session source) throws ProcessRequestException {
                logger.debug("Received DataSm: {}", dataSm);
                return null;
            }
        };
    }

    /**
     * Get human-readable error text for SMPP command status.
     */
    private String getSmppErrorText(int commandStatus) {
        return switch (commandStatus) {
            case 0 -> "ESME_ROK";
            case 4 -> "ESME_RINVMSGLEN";
            case 5 -> "ESME_RINVCMDLEN";
            case 8 -> "ESME_RINVSRCADR";
            case 0xa -> "ESME_RINVDSTADR";
            case 0xb -> "ESME_RINVMSGID";
            case 0xc -> "ESME_RBINDFAIL";
            case 0xd -> "ESME_RINVPASWD";
            case 0xe -> "ESME_RINVSYSID";
            case 0x400 -> "ESME_RTHROTTLED";
            case 0x402 -> "ESME_RINVEXPIRY";
            default -> "command_status=" + commandStatus;
        };
    }
}
